pkg/virt-handler/cmd-client/client.go

```diff
@@ -591,3 +643,58 @@ func (c *VirtLauncherClient) GetFilesystems() (v1.VirtualMachineInstanceFileSyst
 
 	return filesystemList, nil
 }
+
+type ClientFactory struct {
+	clients   map[string]LauncherClient
+	lock      sync.Mutex
+	eventChan chan watch.Event
+	recorder  common.KubernetesEventRecorderInterface
+}
+
+func NewClientFactory(eventChan chan watch.Event, recorderInterface common.KubernetesEventRecorderInterface) *ClientFactory {
+	return &ClientFactory{
+		clients:   map[string]LauncherClient{},
+		lock:      sync.Mutex{},
+		eventChan: eventChan,
+		recorder:  recorderInterface,
+	}
+}
+
+func (f *ClientFactory) EventChan() chan watch.Event {
+	return f.eventChan
+}
+
+func (f *ClientFactory) ClientForSocket(socket string) (LauncherClient, error) {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	if client, exists := f.clients[socket]; exists {
+		return client, nil
+	}
+	client, err := NewClient(socket, f.eventChan, f.recorder)
+	if err != nil {
+		return nil, err
+	}
+	if !IsLegacySocket(socket) {
+		go func() {
+			if err := client.HandleDomainEvents(); err != nil {
```

Comment:
we don't want the client to be created if we can't establish these streams. As the logic is now with this in a goroutine, the stream could fail, but we'd still return a client. I don't know how likely or possible it is, but seems like the error shoudl be handled.

Reply:
Yes, I need to  look a little bit closer into the connection context handling and also need to handle re-connects.